Add benchmarks for rustdoc

jyn514 opened the pull request:
This runs rustdoc by default and requires that you set rustdoc for every perf run. This is probably not the desired behavior, I can change it once everything is working.

Closes https://github.com/rust-lang/rustc-perf/issues/673
Status: Waiting on review

Example usage:
```
$ cargo run --release --bin collector -- --filter hyper --db doc.db bench_local --rustc $RUSTUP_HOME/toolchains/stage1/bin/rustc --rustdoc $RUSTUP_HOME/toolchains/stage1/bin/rustdoc --cargo $RUSTUP_HOME/toolchains/nightly-x86_64-unknown-linux-gnu/bin/cargo no-loops --builds Doc
```

I got it to show up on the site for `bench_commit` but not for `bench_local`. I'm not sure where in the code to look. Suggestions welcome.

![image](https://user-images.githubusercontent.com/23638587/85924439-490a3b80-b860-11ea-9d74-9509e70623e5.png)
![image](https://user-images.githubusercontent.com/23638587/85924445-5e7f6580-b860-11ea-908c-2f548c6e2591.png)

The way this works is by having `rustc-fake` also pretend to be rustdoc if called as `rustdoc-fake`. Additionally there is some hacking around to make the subcommand `rustdoc` instead of `rustc` since currently the subcommand can only be set by a `Profiler`, not by a `BuildKind`.

r? @Mark-Simulacrum

jyn514 commented:
I already noticed that this is doing a lot more work than expected. I thought since rustdoc doesn't check item-bodies it would be faster than `check`, but in some cases it's almost twice as slow. Might be because it has to output the HTML.

The stats also show that rustdoc isn't helped at all by incremental. Not sure if this is an issue with the benchmark or with rustdoc itself. `instructions:u`:

![image](https://user-images.githubusercontent.com/23638587/85924566-33494600-b861-11ea-93b6-25ba90ef7b98.png)

jyn514 commented:
The CI failure looks unrelated to this change, maybe a regression in nightly?

jyn514 commented:
Getting `thread 'tokio-runtime-worker' panicked at 'InvalidColumnType(0, "min(value)", Null): series=46, aid=ArtifactIdNumber(2)', database/src/pool/sqlite.rs:336:37` when I try to compare it to other commits. Not sure why, `select * from pstat where series=46;` and `select * from pstat where aid=2;` both have output but not the two combined.

Mark-Simulacrum commented:
Heh, turns out CI here is failing due to an upstream soundness fix, though not one that affected us (relevant code was a panic! function anyway). Fixed in 72d38b62a9348b041020409a8cf128053b7cb7b3, and closing/reopening to hopefully re-trigger CI.

I'll take a look at your questions here in a bit.

Mark-Simulacrum commented on collector/src/bin/rustc-fake.rs:
```diff
@@ -4,8 +4,18 @@ use std::process::Command;
+    let mut args = args_os.collect::<Vec<_>>();
     let rustc = env::var_os("RUSTC_REAL").unwrap();
+    let rustdoc = env::var_os("RUSTDOC_REAL").unwrap();
+    let actually_rustdoc = dbg!(name).ends_with("rustdoc-fake");
```
nit: leftover `dbg!`

Mark-Simulacrum commented on collector/src/bin/rustc-perf-collector/execute.rs:
```diff
@@ -276,7 +281,11 @@ impl<'a> CargoProcess<'a> {
 
-                profiler.subcommand()
+                if self.build_kind == BuildKind::Doc {
+                    "doc"
```
Hm, this seems like the wrong subcommand -- we'll probably want `rustdoc` not `doc` as we want the same behavior of checking just the last crate (and not all deps).

jyn514 replied:
Ooh, I didn't know `rustdoc` was a thing! That will let me get rid of the `RUSTFLAGS` hack too :)

Mark-Simulacrum commented on collector/src/bin/rustc-perf-collector/main.rs:
```diff
@@ -502,6 +521,23 @@ fn main_result() -> anyhow::Result<i32> {
+            let rustc = which("rustc")?;
+            let rustdoc = which("rustdoc")?;
+            let cargo = which("cargo")?;
+            /*
```
nit: mind removing the comments here?

jyn514 commented:
https://github.com/rust-lang/rustc-perf/pull/675#issuecomment-650574894 was https://github.com/rusqlite/rusqlite/issues/777

jyn514 commented:
Still not sure how to get `bench_local` results to show up on the site. Other than that things seem to be working though.

Mark-Simulacrum commented:
You mean on the graph pages? That's just not possible since we don't have an ordering between them. I guess we could plausibly record the time they were benchmarked -- but that seems like something for a future PR.

Mark-Simulacrum commented on collector/src/bin/rustc-perf-collector/execute.rs:
```diff
@@ -276,7 +281,11 @@ impl<'a> CargoProcess<'a> {
+                    "rustdoc"
+                } else {
+                    profiler.subcommand()
+                }
```
This feels a bit not nice. Can we instead pass in the build kind, and match on that within the subcommand impls? We'd assert that the kind is debug/opt for the llvm-lines subcommand, I guess.

jyn514 replied:
>  We'd assert that the kind is debug/opt for the llvm-lines subcommand, I guess.

We already pass in the build kind for `is_build_kind_allowed`. Maybe we could combine that and `subcommand()`, so that subcommand now returns an Option? That would avoid asserts.

Mark-Simulacrum replied:
Sure, that works too.

Mark-Simulacrum commented on database/src/pool/sqlite.rs:
```diff
@@ -332,7 +333,12 @@ impl Connection for SqliteConnection {
+                                    if let rusqlite::Error::InvalidColumnType(_, _, rusqlite::types::Type::Null) = e {
+                                        None
+                                    } else {
+                                        panic!("{:?}: series={:?}, aid={:?}", e, sid, cid);
```
Looking at that issue, it sounds like the way to do this is to query an Option<T> rather than T (i.e., adjust query_row's inferred return type to `rusqlite::Result<Option<T>>` or so).

Mark-Simulacrum commented:
I guess this doesn't touch the UI at all - in particular, the primary graphs page still has 3 columns. But I think it makes sense for now to keep it that way. We might want to show rustdoc performance on the dashboard, but that can wait several release cycles (until we have a beta and such that also has that data).

jyn514 commented:
> You mean on the graph pages?

I meant on any page, including the compare page. Strangely it seems to be working now, maybe I had a typo in the ID before (??) The stats are all over the place though, I'm not sure what's wrong but I can't believe there was an 80% increase in instructions from a PR that only touched rustc_resolve.

![image](https://user-images.githubusercontent.com/23638587/86063361-b40e6a80-ba38-11ea-9047-03b5f974c65c.png)

Mark-Simulacrum commented:
That looks like you're comparing an official run -- from perf.rust-lang.org's collection server -- to a local run. That won't work well -- the official run, amongst other things, is built with a more optimally configured rustc than is typical for local building. In particular, it has LLVM ThinLTO enabled, and no debug assertions etc.

jyn514 commented:
Ooh, that looks way better! 
![image](https://user-images.githubusercontent.com/23638587/86069919-4dde1380-ba49-11ea-92ee-5c9f26cf968b.png)

Thanks for the help :) I addressed the review comments too.

Mark-Simulacrum commented:
Okay, having done some more thinking here, I think that before we merge this in I'd like to do some work to move [the initial build](https://github.com/rust-lang/rustc-perf/blob/c790398129e5550ae48c0af450dfb78f89eb03c6/collector/src/bin/rustc-perf-collector/execute.rs#L958-L964) outside of each build kind, to cache build dependencies between check/debug/opt/doc builds. For some crates that do codegen etc through build.rs that could be significant. https://github.com/rust-lang/rustc-perf/pull/678 should implement that; I plan to merge it once CI is passing.

It would be great if you could also rebase this on master once that PR merges and add doc builds to CI -- should be fairly easy, adjusting this line: https://github.com/rust-lang/rustc-perf/blob/c790398129e5550ae48c0af450dfb78f89eb03c6/collector/src/bin/rustc-perf-collector/main.rs#L636. I suspect some of the crates will run into trouble, though can't be certain -- maybe things will just work out.

jyn514 commented:
I'm running `bench_local --builds Doc --runs Full` locally to see if any problems pop up, if I haven't reported back in an hour feel free to bump me to post the results.

jyn514 commented:
It's been stuck on `Running coercions: Doc + [Full]` for an hour, not sure what the issue is. I'll try rebasing but I doubt it will help.

Mark-Simulacrum commented:
> It's been stuck on Running coercions: Doc + [Full] for an hour, not sure what the issue is. I'll try rebasing but I doubt it will help.

Hm, that's odd. `cargo doc` in that directory completed in 3 seconds for me. Maybe try upping to RUST_LOG=collector=trace? That way at least you'll know the exact command being run and can hopefully reproduce outside.

jyn514 commented:
Ok, the error is

```
[2020-07-02T18:19:33Z WARN  collector::execute] failed to deserialize stats, retrying; output: Output { status: ExitStatus(ExitStatus(0)), stdout: "", stderr: " Documenting issue-32278-big-array-of-strings v0.1.0 (/tmp/.tmpPEFu6R)\n[collector/src/bin/rustc-fake.rs:13] name = \"/home/joshua/.local/lib/cargo/target/release/rustdoc-fake\"\n    Finished dev [unoptimized + debuginfo] target(s) in 4.22s\n" }
```

So I just have some debug logging gone rogue, oops. The real bug is that it will retry infinitely many times.

jyn514 commented:
And a secondary error: Since I used hard links and not symbolic links, `rustdoc-fake` will not be updated on changes to `rustc-fake`. I'll fix that.

jyn514 commented:
Removing the logging did not fix the issue. I think `MeasureProcessor` is depending on some output that's different between rustdoc and rustc. I'm not sure why this didn't show up until now.

Mark-Simulacrum commented:
My guess is that we're not picking up the rustdoc run as special perhaps, and not dumping perf stat output as such?

jyn514 commented:
When I run rustdoc-fake without the collector things work fine:

```
RUSTC_REAL=$(which rustc) RUSTDOC_REAL=$(which rustdoc) $CARGO_TARGET_DIR/release/rustdoc-fake --wrap-rustc-with perf-stat ~/src/test-rustdoc/primitive-anchor.rs 
2075987279;;instructions:u;466317390;100.00;1.35;insn per cycle
1539796372;;cycles:u;466317390;100.00;3.302;GHz
466.29;msec;task-clock:u;466285639;100.00;0.959;CPUs utilized
466.28;msec;cpu-clock:u;466285639;100.00;0.959;CPUs utilized
28773;;faults:u;466285639;100.00;0.062;M/sec
207432;;max-rss;3;100.00
0.556541778;;wall-time;4;100.00
```

I'm not sure what's going on.

jyn514 commented:
Ok, we're telling cargo to pass `--wrap-rustc-with`, but it never gets passed.

```
[2020-07-02T19:41:00Z DEBUG collector::execute] "/home/joshua/.local/lib/rustup/toolchains/nightly-x86_64-unknown-linux-gnu/bin/cargo" "rustdoc" "--manifest-path" "Cargo.toml" "-p" "file:///tmp/.tmp7c3bbt#issue-32278-big-array-of-strings:0.1.0" "--" "--wrap-rustc-with" "perf-stat"
[collector/src/bin/rustc-fake.rs:10] args_os.collect::<Vec<_>>() = [
    "--crate-type",
    "bin",
    "--crate-name",
    "issue_32278_big_array_of_strings",
    "src/main.rs",
    "-o",
    "/tmp/.tmp7c3bbt/target/doc",
    "--error-format=json",
    "--json=diagnostic-rendered-ansi",
    "--document-private-items",
    "-L",
    "dependency=/tmp/.tmp7c3bbt/target/debug/deps",
]
```

jyn514 commented:
Ugh, I can't replicate at all without going through collector. I'm out of ideas.

jyn514 commented:
This is a bug in cargo. `cargo rustdoc` does not pass the arguments if documenting a binary 🤦

jyn514 commented:
https://github.com/rust-lang/cargo/issues/8445

Mark-Simulacrum commented:
We should be able to make it a library if that helps, FWIW, I don't think it being a binary has any impact on it's timing (or at least I'd not expect so).

jyn514 commented:
In the meantime I'll fix the infinite loop and we just won't have stats for `coercions`. Does that sound reasonable? Or do you want to wait until the cargo bug is fixed?

jyn514 commented:
There are 29 remaining binaries. I'd rather not fix them all one by one, I think it would be better for the cargo bug to get fixed since we're already using nightly.

Mark-Simulacrum commented:
So many! I didn't think we had that many actually being benchmarked. I agree that given we have that many we should probably revert the coercions change and wait for a Cargo fix.

jyn514 commented:
I ran `bench_local --builds Doc --runs Full` with the cargo fix (https://github.com/rust-lang/cargo/pull/8449) and it finished successfully. However the site refused to start up, I think because there were no `Check` runs? I don't think it will impact anything if Doc is run alongside other performance measurements.

```
$ cargo +nightly run --release --bin site -- --db tmp.db
Starting server with port=2346
Loading complete but no data identified; exiting.
```

nnethercote commented:
To start the site I just run `./target/release/site $DATABASE`. I.e. no `--` or `--db` arguments.

jyn514 commented:
There we go!

![image](https://user-images.githubusercontent.com/23638587/86852917-9027d600-c083-11ea-88e4-bc7e2c30786b.png)

I'm not sure how often cargo publishes ... I updated `nightly` this morning but I'm still seeing `cargo 1.46.0-nightly (fede83ccf 2020-07-02)`.

jyn514 commented:
The cargo update landed but hasn't been distributed on nightly yet. So this should be good to go when that happens sometime tomorrow.

Manishearth commented:
@jyn514 does rustc-perf use nightly cargo or the cargo in each master release?

jyn514 commented:
@Manishearth nightly

jyn514 commented:
Same error as on https://github.com/rust-lang/rustc-perf/pull/680#issuecomment-654501953.

Mark-Simulacrum commented:
Okay, that should have triggered CI to run again which should be fixed now. I'll hopefully be merging this soon after that passes (if I haven't feel free to ping me).

Mark-Simulacrum commented:
@jyn514 Actually, could we revert the coercions change from binary to library now that the Cargo patch has landed? I'd prefer to not land that together with this since it could affect perf in theory.

jyn514 commented:
> @jyn514 Actually, could we revert the coercions change from binary to library now that the Cargo patch has landed? I'd prefer to not land that together with this since it could affect perf in theory.

Oops, good catch. Should be reverted now.

jyn514 commented:
@Mark-Simulacrum mind looking at this when you get a chance? I'd love to land this before merging https://github.com/rust-lang/rust/pull/73566 so we have something to compare that to.
